eth/consensus/clique/encoding.py

```diff
@@ -0,0 +1,120 @@
+import rlp
+from typing import (
+    Tuple,
+)
+
+from eth_typing import (
+    Address,
+)
+
+from eth.rlp.sedes import (
+    uint256,
+)
+from eth.consensus.clique.datatypes import (
+    Snapshot,
+    Tally,
+    Vote,
+    VoteAction,
+)
+
+
+def encode_address_tally_pair(pair: Tuple[Address, Tally]) -> bytes:
+    return rlp.encode([
+        rlp.encode(pair[0], sedes=rlp.sedes.binary),
+        rlp.encode(encode_tally(pair[1]), sedes=rlp.sedes.binary)],
+        sedes=rlp.sedes.CountableList(rlp.sedes.binary)
+    )
+
+
+def decode_address_tally_pair(vote: bytes) -> Tuple[Address, Tally]:
+
+    container = rlp.decode(vote, sedes=rlp.sedes.CountableList(rlp.sedes.binary))
+    address = rlp.decode(container[0], sedes=rlp.sedes.binary)
+    tally = decode_tally(rlp.decode(container[1], sedes=rlp.sedes.binary))
+
+    return address, tally
+
+
+def encode_vote(vote: Vote) -> bytes:
+    return rlp.encode([
+        rlp.encode(vote.signer, sedes=rlp.sedes.binary),
+        rlp.encode(vote.block_number, sedes=uint256),
+        rlp.encode(vote.subject, sedes=rlp.sedes.binary),
+        rlp.encode(vote.action.value, sedes=rlp.sedes.binary)],
+        sedes=rlp.sedes.CountableList(rlp.sedes.binary)
+    )
+
+
+def decode_vote(vote: bytes) -> Vote:
+
+    container = rlp.decode(vote, sedes=rlp.sedes.CountableList(rlp.sedes.binary))
+    signer = rlp.decode(container[0], sedes=rlp.sedes.binary)
+    block_number = rlp.decode(container[1], sedes=uint256)
+    subject = rlp.decode(container[2], sedes=rlp.sedes.binary)
+    action = rlp.decode(container[0], sedes=rlp.sedes.binary)
```

Comment:
is the `0` index here repeated on purpose?

Also, do you know about using `rlp.decode(some_bytes, sedes.List((sedes.binary, sedes.big_endian_int)))` to decode something into the rich values in one-go.

Reply:
Uh, no that is a bug. Notice that I don't actually use the rlp encoding for the snapshots that much yet which is probably the reason why this went unnoticed so far.

>Also, do you know about using rlp.decode(some_bytes, sedes.List((sedes.binary, sedes.big_endian_int))) to decode something into the rich values in one-go.

Sweet, I did not!